File: src/main/java/org/interledger/everledger/util/AuthManager.java
package org.interledger.everledger.util;


import java.util.Base64;
import java.util.HashMap;
import java.util.Map;





//import io.vertx.ext.auth.AuthProvider;
import io.vertx.core.http.HttpHeaders;
import io.vertx.core.http.HttpServerRequest;
import io.vertx.ext.web.RoutingContext;
//import io.vertx.ext.web.handler.AuthHandler;





import org.interledger.everledger.AuthInfo;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

public class AuthManager {

    private static final Logger log = LoggerFactory.getLogger(AuthManager.class);
    
//    private final AuthProvider authProvider = new -();
//    String realm = authConfig.getString(DEFAULT_BASIC_REALM, Auth.realm);
//    public final AuthHandler authHandler = BasicAuthHandler.create(authProvider, realm);

    private static final Map<String, AuthInfo> users = new HashMap<String, AuthInfo>();

    public static Map<AuthInfo, Integer /*blance*/> configureDevelopmentEnvironment() {
        if (! org.interledger.everledger.Config.unitTestsActive) {
            throw new RuntimeException("developer.unitTestsActive must be true @ application.conf "
                    + "to be able to reset tests");
        }
        Map<AuthInfo, Integer /*blance*/> result = new HashMap<AuthInfo, Integer /*blance*/>();
        AuthInfo admin        = new AuthInfo(        "admin",        "admin",        "admin", "admin");
        AuthInfo ilpconnector = new AuthInfo( "ilpconnector", "ilpconnector", "ilpconnector", "connector");
        AuthInfo alice        = new AuthInfo(        "alice",        "alice",        "alice", "user");
        AuthInfo bob          = new AuthInfo(          "bob",          "bob",          "bob", "user");
        // AuthInfo noBalance    = new AuthInfo(    "nobalance",    "nobalance",    "nobalance", "user"); 
        AuthInfo eve          = new AuthInfo(          "eve",          "eve",          "eve", "user");

        users.put("admin"       , admin       ); result.put(admin       , 10000);
        users.put("ilpconnector", ilpconnector); result.put(ilpconnector,   100);
        users.put("alice"       , alice       ); result.put(alice       ,   100);
        users.put("bob"         , bob         ); result.put(bob         ,     0);
        users.put("eve"         , eve         ); result.put(eve         ,     0);
        // users.put("nobalance"   , noBalance   ); result.put(noBalance   ,     0);
        return result;
    }
    
    public static Map<String, AuthInfo> getUsers() {
        return users;
    }
    
    public static void setUser(String id, String pass, String roll){
        final AuthInfo ai = new AuthInfo(id, id, pass, roll);
        users.put(id, ai);
    }

    public static AuthInfo authenticate(RoutingContext context, boolean allowAnonymous) {
        HttpServerRequest request = context.request();
        String authorization = request.headers().get(HttpHeaders.AUTHORIZATION);
        if (authorization == null) {
            if (allowAnonymous) { return AuthInfo.ANONYMOUS; }
            throw ILPExceptionSupport.createILPUnauthorizedException();
        }
        try {
            String[] parts = authorization.split(" ");
            String sscheme = parts[0];
            if (!"Basic".equals(sscheme)) {
                log.error("Only Basic Authorization support supported.");
                throw ILPExceptionSupport.createILPUnauthorizedException();
            }
            String decoded = new String(Base64.getDecoder().decode(parts[1]));
            int colonIdx = decoded.indexOf(":");
            String suser = (colonIdx != -1) ? decoded.substring(0, colonIdx) : decoded;
            String spass = (colonIdx != -1) ? decoded.substring(colonIdx + 1): null ;
            AuthInfo authInfo = users.get(suser);
            if (authInfo == null){
                log.error("authInfo null. (User not int AuthManager.users lists)");
                throw ILPExceptionSupport.createILPUnauthorizedException();
            }
            boolean isAdmin = authInfo.getRoll().equals("admin");
            if (!isAdmin && !authInfo.pass.equals(spass)  ) {
                log.error("user "+authInfo.id+" is not admin and pass doesn't match");
                throw ILPExceptionSupport.createILPUnauthorizedException();
            }
            return authInfo;
        } catch (Exception e) {
            log.error("Unhandled Auth Exception: " + e.toString());
            throw ILPExceptionSupport.createILPUnauthorizedException();
        }
    }

    public static AuthInfo authenticate(RoutingContext context) {
        return authenticate(context, false /*do not allow anonymous access */);
    }

}

File: src/main/java/com/everis/everledger/util/AuthManager.kt
package com.everis.everledger.util

import com.everis.everledger.AccessRoll
import com.everis.everledger.AuthInfo
import com.everis.everledger.util.Config
import io.jsonwebtoken.Jwts
import io.jsonwebtoken.SignatureAlgorithm
import io.jsonwebtoken.impl.crypto.MacProvider
import io.vertx.core.http.HttpHeaders
import io.vertx.ext.web.RoutingContext
import org.slf4j.LoggerFactory
import java.security.Key
import java.util.*

object AuthManager {

    private val log = LoggerFactory.getLogger(AuthManager::class.java)
    var SigAlgth = SignatureAlgorithm.HS256
    var key: Key = MacProvider.generateKey(SigAlgth)
    var parser = Jwts.parser().setSigningKey(key)


    //    private final AuthProvider authProvider = new -();
    //    String realm = authConfig.getString(DEFAULT_BASIC_REALM, Auth.realm);
    //    public final AuthHandler authHandler = BasicAuthHandler.create(authProvider, realm);

    private var users: MutableMap<String, AuthInfo> = HashMap()

    fun configureDevelopmentEnvironment(): Map<AuthInfo, IntArray/*[balance,minAllowedBalance]*/> {
        // TODO:(0) This is a temporal "patch". Ideally functional-test will recreate the accounts
        //   through HTTP-requests.
        if (!Config.unitTestsActive) {
            throw RuntimeException("developer.unitTestsActive must be true @ application.conf " + "to be able to reset tests")
        }
        val result = HashMap<AuthInfo, IntArray/*balance*/>()
        val admin = AuthInfo(Config.test_ethereum_address_admin, "admin", "admin", AccessRoll.ADMIN)
        val ilpconnector = AuthInfo(Config.test_ethereum_address_ilpconnector, "ilpconnector", "ilpconnector", AccessRoll.CONNECTOR)
        val alice = AuthInfo(Config.test_ethereum_address_alice, "alice", "alice", AccessRoll.USER)
        val bob = AuthInfo(Config.test_ethereum_address_bob, "bob", "bob", AccessRoll.USER)
        // AuthInfo noBalance    = new AuthInfo(    "nobalance",    "nobalance",    "nobalance", "user");
        val eve = AuthInfo(Config.test_ethereum_address_eve, "eve", "eve", AccessRoll.USER)
        users = HashMap<String, AuthInfo>()
        users.put(admin.login, admin)
        result.put(admin, intArrayOf(10000,-1000000000))
        users.put(ilpconnector.login, ilpconnector)
        result.put(ilpconnector, intArrayOf(100,0))
        users.put(alice.login, alice)
        result.put(alice, intArrayOf(100,0))
        users.put(bob.login, bob)
        result.put(bob, intArrayOf(0,0))
        // users.put("nobalance"   , noBalance   ); result.put(noBalance   ,     0);
        return result
    }

    fun getUsers(): Map<String, AuthInfo> {
        return users
    }

    fun setUser(ai: AuthInfo) {
        users.put(ai.login, ai)
    }

    @JvmOverloads fun authenticate(context: RoutingContext, allowAnonymous: Boolean = false): AuthInfo {

            val request = context.request()

            //  "GET /websocket?token=eyJhbGciOiJIUzUxMiJ9.eyJzdWIiOiJodHRwOi8vMTcyLjE3LjAuMTozMDAxLy9hY2NvdW50cy9hZG1pbiIsImlzcyI6IjE3Mi4xNy4wLjEifQ.yt95JiNCOzwn80MVP25KfXNhyHfxZmiclwCFATSN7wVNm3ODazPmaEqf8TLkvkiDJoHM49LqvDzgvzbKe_rSxw
            val token = request.getParam("token")
            if (token != null) {// REF: ./services/auth.js
                if (!parser.isSigned(token)) {
                    throw ILPExceptionSupport.createILPUnauthorizedException()
                }
                val URLAccount = parser.parseClaimsJws(token).body.subject
                val offset = URLAccount.lastIndexOf("/accounts/")
                val suser = URLAccount.substring(offset + "/accounts/".length)
                val authInfo = users[suser]
                if (authInfo == null) {
                    log.error("authInfo null. (User not int AuthManager.users lists)")
                    throw ILPExceptionSupport.createILPUnauthorizedException()
                }
                return authInfo
            }
            val authorization : String
            try {
                authorization = request.headers().get(HttpHeaders.AUTHORIZATION)
            }catch(e : Exception){
                if (allowAnonymous) return AuthInfo.ANONYMOUS
                throw ILPExceptionSupport.createILPUnauthorizedException()
            }
            if (authorization == null) {
                if (allowAnonymous) return AuthInfo.ANONYMOUS
                else throw ILPExceptionSupport.createILPUnauthorizedException()
            }
            val parts = authorization.split(" ".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
            val sscheme = parts[0]
            if ("Basic" != sscheme) {
                log.error("Only Basic Authorization support supported.")
                throw ILPExceptionSupport.createILPUnauthorizedException()
            }
            val decoded = String(Base64.getDecoder().decode(parts[1]))
            val colonIdx = decoded.indexOf(":")
            val suser = if (colonIdx != -1) decoded.substring(0, colonIdx) else decoded
            val spass = if (colonIdx != -1) decoded.substring(colonIdx + 1) else null

            val authInfo = users[suser]
            if (authInfo == null) {
                log.error("users[" + suser + "] null. (User not int AuthManager.users lists)")
                throw ILPExceptionSupport.createILPUnauthorizedException()
            }
            if (!authInfo.isAdmin && authInfo.pass != spass) {
                log.error("user " + authInfo.id + " is not admin and pass doesn't match")
                throw ILPExceptionSupport.createILPUnauthorizedException()
            }
            return authInfo
    }

}/*do not allow anonymous access */